snapcraft/internal/meta/_snap_packaging.py

```diff
@@ -133,6 +133,97 @@ def _update_yaml_with_extracted_metadata(
         raise meta_errors.MissingSnapcraftYamlKeysError(keys=missing_keys)
 
 
+def _adopt_info(
+        config_data: Dict[str, Any],
+        extracted_metadata: _metadata.ExtractedMetadata):
+    metadata_dict = extracted_metadata.to_dict()
+    for key, value in filter(
+            lambda item: item != 'desktop',
+            metadata_dict.items()):
+        if key not in config_data:
+            if key == 'icon':
+                if _icon_file_exists() or not os.path.exists(value):
+                    # Do not overwrite the icon file.
+                    continue
+            config_data[key] = value
+    if 'desktop' in metadata_dict:
+        for desktop_id in metadata_dict['desktop']:
+            app_name = _get_app_name_from_desktop_id(config_data, desktop_id)
+            if app_name and not _desktop_file_exists(app_name):
+                for desktop_file_path in (
+                        os.path.join(
+                            xdg_data_dir, 'applications',
+                            desktop_id.replace('-', '/')) for
+                        xdg_data_dir in ('usr/local/share', 'usr/share')):
+                    if os.path.exists(desktop_file_path):
+                        config_data['apps'][app_name]['desktop'] = (
+                            desktop_file_path)
+
+
+def _icon_file_exists() -> bool:
+    """Check if the icon is specified as a file in the assets dir.
+
+    The icon file can be in two directories: 'setup/gui' (deprecated) or
+    'snap/gui'. The icon can be named 'icon.png' or 'icon.svg'.
+
+    :returns: True if the icon file exists, False otherwise.
+    """
+    for icon_path in (
+            os.path.join(asset_dir, 'gui', icon_file) for
+            (asset_dir, icon_file) in itertools.product(
+                ['setup', 'snap'],
+                ['icon.png', 'icon.svg'])):
+        if os.path.exists(icon_path):
+            return True
+    else:
+        return False
+
+
+def _get_app_name_from_desktop_id(
+        config_data: Dict[str, Any], desktop_id: str) -> str:
```

Comment:
Hmm... yeah I don't feel like this scales well. The reason we have to do this is obvious: apps aren't tied to parts. However, I have two thoughts:

1) Why use the file name? Why not actually take a look at the `Exec` line to see what app it's running? We don't want to ship desktop files that don't run an app in the snap, anyway.

2) App-specific `desktop` entries are only _one_ way to get .desktop files into the snap. The other way is the conventional location: the `meta/gui` directory. What if we just installed all the adopted desktop files there, no YAML updates needed? Then we could take a look at the `Exec` lines and filter based on our list of app names. Thoughts?

Reply:
The Exec would work, I think. But would you expect the app name or the command in the Exec?
I don't like very much because we rewrite the Exec line during prime, so it feels more like a throwaway value to me. Also, snaps sometimes force us to have an app name that's not the same as the executed command. 

I'm not sure which is best, but I would prefer to keep it simple mostly because this is temporary. Once we have the appstream-id there is no more reason to guess.

About your second point, I'm not sure I get it. The other way to define them is to put them in `snap/gui`, and they are matched with the corresponding app by file name. They have to be called test-app.desktop. And during prime they are moved to `meta/gui`. So do you want to replace matching them by name to match them by exec?

Reply:
> I don't like very much because we rewrite the Exec line during prime, so it feels more like a throwaway value to me.

Good call, I forgot about that.

> About your second point, I'm not sure I get it. The other way to define them is to put them in snap/gui, and they are matched with the corresponding app by file name.

I don't see them being matched anywhere, so perhaps we're misunderstanding each other. They're just [migrated straight over](https://github.com/snapcore/snapcraft/blob/master/snapcraft/internal/meta/_snap_packaging.py#L229), and we can put stuff there ourselves as well. I don't think we need to rename them at all, but they do need to have a valid Exec line, which is perhaps too much to ask for.

Reply:
> I'm not sure which is best, but I would prefer to keep it simple mostly because this is temporary. Once we have the appstream-id there is no more reason to guess.

But that only stands for appstream metadata, right? What about desktop files coming from something else?

Reply:
Sorry, I was rebasing and apparently did it wrong. I'm happy to revisit this, we can still change the behaviour without much fuss even after this lands. Specially, because what you suggested preserves the behaviour for the most common cases.

So I'm looking at the code that generates the desktop files again, and it's very confusing. I've just noticed that when the desktop keyword is specified in the snapcraft.yaml, the exec line is rewritten. But when the file is saved in snap/gui, it is left untouched. That's inconsistent, and I think it's wrong. I think that when the file is in snap/gui, we should move it to prime/meta/gui, and check if there's an app with the same name as the desktop file. If so, rewrite the exec line to make it consistent with the other way of specifiying them.

This is what I thought that was happening, but it's only for half of the story.

Reply:
About the other question, that's one of the reasons why I wanted to leave all the metadata handling to the prime step only. That way we would have a single place that requires to know the specifics of the metadata source, to extract and convert at the same time.

Once we add appstream id support, we can turn this desktop file ids into a dictionary that maps appstream id to desktop file id. As you point out, this is appstream-specific, so the field in the metadata object should be called appstream_launchable, or something. I see no way around this with the current implementation. I'm happy to meet to talk more about this too.